daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -483,6 +483,132 @@ sealed abstract class HasTxNodes[Nid, +Cid] {
     }
   }
 
+  /** Return the expected contract key inputs (i.e. the state before the transaction)
+    * for this transaction or an error if the transaction contains a
+    * duplicate key error or has an inconsistent mapping for a key.
+    *
+    * Because we do not preserve byKey flags across transaction serialization
+    * this method will consider all operations with keys for conflicts
+    * rather than just by-key operations.
```

Comment:
I am  a bit uncomfortable with the "confusing semantics" argument when dealing with backward compatibility:
"Of course it is confusing, we are dealing with backward compatibility"

The question am I interested by is:  won't a ledger implementer that does not have to deal with version older than minByKey` tempted to implement that for byKey node only, hence defeating the purpose of this generic method ?  

Not saying we should go this way, just trying to challenge you through process.  
If you persist please modify the comment with something like: 
```suggestion
    * Because transaction serialization for LF <= 1.13 does not preserve byKey flags,
    * this method will consider all operations with keys for conflicts
    * rather than just by-key operations.
```

Reply:
> The question am I interested by is: won't a ledger implementer that does not have to deal with version older than minByKey` tempted to implement that for byKey node only, hence defeating the purpose of this generic method ?

I don’t really see why they would. This method is perfectly fine even for newer versions and they should be able to use it safely. At worst, they get some extra checks or inputs that might not be required.

I think what we might want to consider is to have a separate method that _only_ works for LF > 1.13 and does implement the `byKey` semantics. The nice thing about this is that at that point we will never get `Left` since the checks match up with Speedy.

What I’d like to avoid is mixing the semantics. That gets pretty confusing and like an easy way to introduce bugs.

At worst, you might end up